### DISCUSSION TITLE
Weird behavior (ICMP not intercepted)

### DISCUSSION THREAD

**Discussion by akronym0** [evilsocket/opensnitch#946] [Q&A] [answered]
First of all ... I love OpenSnitch. Thanks for that! Now I am a full-time Linux desktop user :-)

I have been getting "pop up dialogs" about apps unexpectedly connecting to somewhere. Unexpectedly, because there had been no reason for the app to connect to the internet, but I have been thinking, well, that's why we need application firewalls. "Good job, OpenSnitch!".

But then I noticed that those apps tried to connect to the same sites that I was browsing with Firefox at the same time. Let's say, I was browsing news.xyz, then sometimes OpenSnitch warned that an unrelated app tried to ping news.xyz as well.

Is there an explanation for this weird behavior?

**Comment by gustavo-iniguez-goya** [comment]
hi @akronym0 !

A "couple" of questions O:)

What opensnitch version are you using? and what process monitor method? You can set LogLevel to DEBUG under Preferences->Nodes, and see if is there any errors using the `ebpf` process monitor method. 
If everything goes fine, you should see a message in `/var/log/opensnitchd.log` like this one :point_right: ` INF  Process monitor method ebpf`

Do you suspend your computer?
If you could post a domain+ip that you load on firefox and is attributed to other app it'd help to debug this issue.

> there had been no reason for the app to connect to the internet

Do these apps connect to the internet on a regular basis? or not at all?

**Comment by akronym0** [comment]
Thanks for your quick response.

I just the set log level to debug. And I am waiting for it to happen again. I am not able to reproduce this at will yet :-) 

I am on version 1.6rc5,

Those apps  are not expected to connect to the internet, when they are not in use. I hope that makes sense. But I suspect they did not actually try to connect. There must have been some kind of "interference" with Firefox (I doubt those apps would have been able to "see" what sites I opened with Firefox). But for some reason OpenSnitch attributed the connection to those apps.

Yes, I frequently hibernate (to disk) and suspend (to memory). But I don't remember whether that is connected. I will keep this in mind while waiting for it to happen again.

**Reply by gustavo-iniguez-goya** [reply]
thank you @akronym0 , try to reproduce it. When the bug hits again, take a look at the logs and look for logs like:
```
[0/5] outgoing connection uid: 1000, 5353:172.17.0.1 -> 224.0.0.251:5353 || netlink response: 5353:224.0.0.251 -> 0.0.0.0:0 inode: 1840325 - loopback: false multicast: false unspecified: true linklocalunicast: false ifaceLocalMulticast: false GlobalUni: false
(...)
[2023-05-21 16:42:39]  DBG  Socket found in known pids 54.483µs, pid: 328848, inode: 2138310401, pos: 0, pids in cache: 31
[2023-05-21 16:42:39]  DBG  [0] PID found 328848 [2138310401]
```

if ebpf is working as expected you should only see logs entries like these ones:
```
[2023-05-21 16:43:26]  DBG  new connection tcp => 38118:172.21.0.5 -> 172.21.0.6 ():7946 uid: 4294967295, mark: 0
[2023-05-21 16:43:26]  DBG  [ebpf conn] not in cache, but in execEvents: tcp38118172.21.0.5172.21.0.67946, 2108627 -> /usr/bin/loki -> [/usr/bin/loki -config.file=/etc/loki/config.yaml -target=read -reporting.enabled=false]
[2023-05-21 16:43:26]  DBG  [ebpf conn] adding item to cache: tcp38118172.21.0.5172.21.0.67946
[2023-05-21 16:43:26]  DBG  ✔ /usr/bin/loki -> 38118:172.21.0.5 => 172.21.0.6:7946, mark: 0
```

I described here a problem after coming back from Suspend (I blamed Gobpf, but it also happens with the Cilium lib): https://github.com/evilsocket/opensnitch/discussions/834#discussioncomment-5712431
You can also use the script posted there to see if any of the ebpf maps (usually tcp) is full (12k elements).

**Comment by akronym0** [comment]
Ok. I downloaded the script and installed the bpftool.

And forgot to answer

> what process monitor method? 

Ebpf

**Comment by akronym0** [comment]
I found an example in the log where a new connection is attributed to the wrong app

```
new connection icmp => 0:10.8.12.26 -> 35.215.92.129 (turn3.voip.signal.org):0 uid: 4294967295
[0/1] outgoing connection uid: 0, 0:10.8.12.26 -> 35.215.92.129:0 || netlink response: 1:0.0.0.0 -> 0.0.0.0:0 inode: 25321 - loopback: false multicast: false unspecified: true linklocalunicast: false ifaceLocalMulticast: false GlobalUni: false 
GetSocketInfo() invalid: 1:0.0.0.0 -> 0.0.0.0:0
netlink socket not found, adding entry:  0:10.8.12.26 -> 35.215.92.129:0 || 1:0.0.0.0 -> 0.0.0.0:0 inode: 25321 state: close
[0/1] outgoing connection uid: 0, 0:10.8.12.26 -> 35.215.92.129:0 || netlink response: 1:0.0.0.0 -> 0.0.0.0:0 inode: 25321 - loopback: false multicast: false unspecified: true linklocalunicast: false ifaceLocalMulticast: false GlobalUni: false 
GetSocketInfo() invalid: 1:0.0.0.0 -> 0.0.0.0:0
netlink socket not found, adding entry:  0:10.8.12.26 -> 35.215.92.129:0 || 1:0.0.0.0 -> 0.0.0.0:0 inode: 25321 state: close
Socket found in known pids 17.623µs, pid: 4527, inode: 25321, pos: 0, pids in cache: 4
[0] PID found 4527 [25321]
✔ /usr/bin/vmnet-natd -> turn3.voip.signal.org:0
```
Basically OpenSnitch found the wrong PID. In this case the PID belongs to the vm ware player app, but obviously OpenSnitch should have found the PID of the Signal messenger app.

And I noticed that this only happens for ICMP connections.

Hope this helps.

**Reply by gustavo-iniguez-goya** [reply]
Ah, I see, we've had a similar report recently. 

We don't intercept icmp packets at kernel level yet, so relaying on proc is not totally useful. But anyway, the scenario is a bit tricky.
tl;dr: we need to work on it. PRs welcomed :)

On the one hand, netfilter redirects us the icmp packets as follow:
`new connection icmp => 0:192.168.1.104 -> 1.1.1.1 ():0`

However nowadays, instead of icmp, pings are udp packets to port 1025 to avoid requiring CAP_RAW/suid perms: https://lwn.net/Articles/443051/
Here're a couple of great posts on how it works: https://echorand.me/posts/my-own-ping/ - https://terenceli.github.io/%E6%8A%80%E6%9C%AF/2021/02/19/ping-1025

hooking `security_socket_connect`, `connect`, `inet_dgram_connect` or `inet_dgram_send`, we intercept packets of type SOCK_DGRAM (2) and proto IPPROTO_UDP (17), destination port 1025. Thus, we don't find outgoing icmp packets by `icmp 0:x.x.x.x -> y.y.y.y:0`.
Yes, we could determine that type 2 + proto 17 to port 1025 is a icmp, but for example hping3 -1 does not use port 1025... and proto is not 17 nor 1, but 255...

Hooking `security_socket_sendmsg` we receive the expected socket parameters: type SOCK_RAW (3), proto IPPROTO_ICMP (1), but src ip is not set (0.0.0.0) . This already happens with some broadcast packets, and ~could~ works.

Maybe the solution here would be to intercept socket(), save the type and proto, and then check it later on another hook (connect, etc).
Or maybe there's a better and direct way of intercepting these packets correctly.

**Reply by gustavo-iniguez-goya** [reply, marked as answer]
I've been reviewing this problem, and as I wrote here https://github.com/evilsocket/opensnitch/discussions/946#discussioncomment-6035934 , this issue in particular is related to DestinationUnreachable ICMPs (and probably others).

These ICMPs seems to be originated from kernel space and we'll have to investigate how to detect them.

For now, this problem can be circumvented around by adding a rule to allow these ICMPs in particular, from the SystemRules:
![image](https://github.com/evilsocket/opensnitch/assets/2742953/42ac3da3-388d-42c9-987d-4063a58d7b74)

With this rule I no longer receive pop-ups to allow these ICMPs.

**Reply by akronym0** [reply]
Thanks, I will also add this rule. May I ask why accept and not drop/reject? Maybe it doesn't matter?

By the way, your pop-up looks nice. What theme do you use? Mine looks like this

![popup2](https://github.com/evilsocket/opensnitch/assets/86428317/bde2ffb1-c50b-4cfc-8b71-3fa93e7e3f83)

(Please note, screenshot shows IN, but I added with OUT ;-)

Linux Mint 21.1 Cinnamon.

**Reply by gustavo-iniguez-goya** [reply]
> May I ask why accept and not drop/reject? Maybe it doesn't matter?

yeah, in this case I think it won't matter. But with VPNs for example, blocking ICMPs may cause not to discover the proper MTU for the tunnel, causing slodowns in the network.

The theme is the ubuntu's 22.10 default one, Yaru I think is called. Try installing qt-material `pip3 install qt-material` , select a theme from Preferences->UI and see if it looks nicer.

Anyway, the alignment of the icons is really strange, first time I see it. I'll test it on LinuxMint 21.1

**Reply by akronym0** [reply]
Thanks, I had tried qt-material already (light_blue_500), but while the alignment is ok, it kind of looks ugly :-) But I am not complaining. It is perfectly fine and usable. So there is no need to investigate anything. Your time is too valuable for that.

**Reply by gustavo-iniguez-goya** [reply]
I've reproduced thi problem on latest LinuxMint 21.1 after changing Monitor scale to 125%
I'll see if I can easily fix it

**Reply by akronym0** [reply]
Great. My scaling factor is 200% (hidpi monitor).

**Comment by akronym0** [comment]
Ok, no problem. Thanks for looking into this. If handling ICMP is a difficult nut to crack, we can wait. And the problem is not too bad at the moment. I have to dismiss the ICMP pop-up roughly 10 times per day. This is "a price I am willing to pay" to keep using OpenSnitch :-)

Thank you again for all your amazing work!

**Reply by gustavo-iniguez-goya** [reply]
I've got a PoC working. It'd be interesting to see if it works on your computer.
If you can compile the daemon + the ebpf module I can post the changes here, or you're willing to test it I can post the binaries.

**Reply by akronym0** [reply]
Yes, I am willing to test. Please post the binaries.

**Reply by gustavo-iniguez-goya** [reply]
Ok, uploaded.

Before test it:
```bash
# backup previous module
~ $ sudo cp /usr/lib/opensnitchd/ebpf/opensnitch.o /usr/lib/opensnitchd/ebpf/opensnitch.o.orig

# copy new module
~ $ gunzip opensnitch.o.gz
~ $ sudo cp opensnitch.o /usr/lib/opensnitchd/ebpf/opensnitch.o
```

Then stop the daemon (`sudo systemctl stop opensnitch`), and launch the new one from the cli:
```bash
~ $ gunzip opensnitchd.946.gz
~ $ sudo su
~ # ./opensnitchd.946 -rules-path /etc/opensnitchd/rules/
```

There's one case of ICMP that I haven't added, socket type DGRAM (2) + proto IPPROTO_IP (1). Common on old kernels (5.8) as unprivileged user. But I guess you'll be running 6.x, which is common nowadays.

**Comment by akronym0** [comment]
Thanks. Just to clarify. What is the expected behavior now? I guess, no ICMP popups anymore, right? Anything else?

I am on 5.19. I wish I could use a 6.x kernel but 6.x has "hibernation" issues on my hardware unfortunately.

**Reply by akronym0** [reply]
And I am not sure starting Opensnitch from the command line will work, because I am usually only able to reproduce this ICMP after a reboot. Is it possible to just copy the 946 binary over the existing version?

**Comment by akronym0** [comment]
Well, I was able to reproduce this without a reboot. But the new version did not fix the issue.


```
sudo ./opensnitchd.946 -rules-path /etc/opensnitchd/rules/
IMP  Starting opensnitch-daemon v1.6.0
INF  Loading rules from /etc/opensnitchd/rules ...
OK: libnetfiler_queue supports nfq_get_uid
OK: libnetfiler_queue supports nfq_get_uid
found /lib/x86_64-linux-gnu/libc.so.6
```


Here is the log

```
new connection icmp => 0:192.168.1.111 -> 104.22.11.50 (media.printables.com):0 uid:                4294967295, mark: 0
[0/1] outgoing connection uid: 0, 0:192.168.1.111 -> 104.22.11.50:0 || netlink response: 17:0.      0.0.0 -> 0.0.0.0:0 inode: 35622 - loopback: false multicast: false unspecified: true linklocalunicast: false ifaceLocalMulticast: false GlobalUni: false
GetSocketInfo() invalid: 17:0.0.0.0 -> 0.0.0.0:0
netlink socket not found, adding entry:  0:192.168.1.111 -> 104.22.11.50:0 || 17:0.0.0.0 -> 0.      0.0.0:0 inode: 35622 state: close
[0/1] outgoing connection uid: 0, 0:192.168.1.111 -> 104.22.11.50:0 || netlink response: 17:0.      0.0.0 -> 0.0.0.0:0 inode: 35622 - loopback: false multicast: false unspecified: true linklocalunicast: false ifaceLocalMulticast: false GlobalUni: false
GetSocketInfo() invalid: 17:0.0.0.0 -> 0.0.0.0:0
netlink socket not found, adding entry:  0:192.168.1.111 -> 104.22.11.50:0 || 17:0.0.0.0 -> 0.      0.0.0:0 inode: 35622 state: close
Inode found in cache: 5.851µs 2969 35622 35622 192.168.1.111 0 104.22.11.50 0
[0] PID found 2969 [35622]
/usr/lib/ipsec/charon^[[0m -> 0:^[[1m192.168.1.111^[[0m =>              ^[[1mmedia.printables.com (104.22.11.50)^[[0m:0, mark: 0 (^[[31mdeny-until-restart-simple-usr-lib-ipsec-charon-debug-ike-1-debug-knl-1-debug-cfg-0^[[0m)
```

**Reply by gustavo-iniguez-goya** [reply]
The expected behaviour was to intercept ICMP packets correctly O:) but it seems that it's not working, or maybe eBPF method is not working.

> And I am not sure starting Opensnitch from the command line will work, because I am usually only able to reproduce this ICMP after a reboot. Is it possible to just copy the 946 binary over the existing version?

Yes, but it'd be better to launch it from the cli, to discard the suspend/hibernation bug.

It's critical that eBPF works, could you perform the following steps?

- Empty the log /var/log/opensnitchd.log
- Copy the .o file to /usr/lib/opensnitchd/ebpf/
- Start the binary from the cli.
- Verify that the following line is logged to the log file: `INF  [eBPF] module loaded: /usr/lib/opensnitchd/ebpf/opensnitch.o`
- Verify that you have at least this entry in the kprobe_events file (besides 5-6 more):
```bash
~ # cat /sys/kernel/debug/tracing/kprobe_events 
p:kprobes/psecurity_socket_sendmsg security_socket_sendmsg
```

You can also monitor kernel events in real time:
```
~ # cat /sys/kernel/debug/tracing/trace_pipe | grep ping
            ping-3075488 [006] d..31 1058746.704057: bpf_trace_printk: sec_socket_sendmsg, type1: 3, proto: 1, fam: 2
            ping-3075488 [006] d..31 1058746.704064: bpf_trace_printk: sec_socket_sendmsg, type1: 3, dport: 0
            ping-3075488 [006] d..31 1058746.704066: bpf_trace_printk: sec_socket_sendmsg, proto: 1, type: 3, comm: ping
```

Finally, execute a ping from another shell and look for these logs:
```
[2023-05-25 21:04:01]  IMP  ICMP!
[2023-05-25 21:04:01]  IMP  ICMP found via eBPF
```

If eBPF is not working, there should be a log like: `[ebpf conn] PID not found via eBPF, falling back to proc`

**Reply by akronym0** [reply]
I did that and I saw everything you have described except for the last part.

I executed `ping github.com`
```
ping-11635   [012] d..31  5339.365778: bpf_trace_printk: sec_socket_sendmsg, type1: 2, proto: 17, fam: 2
ping-11635   [012] d..31  5339.365781: bpf_trace_printk: sec_socket_sendmsg, NOT ICMP => type1: 2, proto: 17
ping-11635   [012] d..31  5339.365806: bpf_trace_printk: sec_socket_sendmsg, type1: 2, proto: 17, fam: 2
ping-11635   [012] d..31  5339.365807: bpf_trace_printk: sec_socket_sendmsg, NOT ICMP => type1: 2, proto: 17
```



But I didn't find this in the log (I checked visually, but also used grep -i icmp and so on)

```
[2023-05-25 21:04:01]  IMP  ICMP!
[2023-05-25 21:04:01]  IMP  ICMP found via eBPF
```
And I didn't see 

`[ebpf conn] PID not found via eBPF, falling back to proc`

either.

**Comment by gustavo-iniguez-goya** [comment]
> sec_socket_sendmsg, NOT ICMP => type1: 2, proto: 17

:( this is the case I didn't include. That "ping" is in reality a UDP packet to port 1025, to avoid using RAW sockets/privs and not require suid permissions.

Did you realize if other ICMPs are being intercepted? => `tail -f /var/log/opensnitchd.log | grep icmp -A 1`

If it's not working the output would look like:
```
new connection icmp => 0:192.168.1.111 -> 104.22.11.50 (media.printables.com):0 uid:                4294967295, mark: 0
[0/1] outgoing connection uid: 0, 0:192.168.1.111 -> 104.22.11.50:0 || netlink response: 17:0.      0.0.0 -> 0.0.0.0:0 inode: 35622 - loopback: false multicast: false unspecified: true linklocalunicast: false ifaceLocalMulticast: false GlobalUni: false
GetSocketInfo() invalid: 17:0.0.0.0 -> 0.0.0.0:0
```

if it's working:
```
[2023-05-26 11:31:17]  DBG  new connection icmp => 0:192.168.1.234 -> 1.1.1.1 ():0 uid: 0, mark: 0
[2023-05-26 11:31:17]  DBG  [ebpf conn] not in cache, NOR in execEvents: icmp0192.168.1.2341.1.1.10, 3179191 -> /bin/ping -> []
[2023-05-26 11:31:17]  DBG  [ebpf conn] adding item to cache: icmp0192.168.1.2341.1.1.10
```

**Reply by akronym0** [reply]
I guess, it is not working ... Because I only have

```
new connection icmp => 0:10.8.12.26 -> 104.22.10.50 (media.printables.com):0 uid: 4294967295, mark: 0
[2023-05-26 12:27:29]  DBG  [0/1] outgoing connection uid: 0, 0:10.8.12.26 -> 104.22.10.50:0 || netlink response: 17:0.0.0.0 -> 0.0.0.0:0 inode: 23876 - loopback: false multicast: false unspecified: true linklocalunicast: false ifaceLocalMulticast: false GlobalUni: false 
```

**Comment by gustavo-iniguez-goya** [comment]
yeah, it looks like all ICMP packes on your system are UDP packets to port 1025. 
We'll need finer control over the sockets to do this properly, for example intercept sockets creation, save the parameters, and later on in securty_socket_sendmsg or other function, determine if the packet is an ICMP.

Categorizing UDP packets to port 1025 as ICMP is not a good idea, because many people will allow ICMP packets blindly. 

This needs time, for now I'll focus no the next release.

**Comment by akronym0** [comment]
According to this

http://terenceli.github.io/%E6%8A%80%E6%9C%AF/2021/02/19/ping-1025

ping does not send udp packets to 1025. It is just a trick to get the current source IP that ping used. If we specify the source IP, no udp will be involved.

So I tried to use this to generate a pure ping

` ping -I 192.168.1.111 1.1.1.1 -c 1 `

But this seems to bypass OpenSnitch. There is no reaction. And there are no log lines that contain icmp. Any ideas why there is no pop-up for that?

**Comment by gustavo-iniguez-goya** [comment]
```
[2023-05-28 11:05:23]  DBG  new connection icmp => 0:192.168.1.234 -> 1.1.1.1 ():0 uid: 0, mark: 0
[2023-05-28 11:05:23]  DBG  [ebpf conn] in cache: icmp0192.168.1.2341.1.1.10, 3614361 -> /bin/ping
[2023-05-28 11:05:23]  DBG  ✔ /bin/ping -> 0:192.168.1.234 => 1.1.1.1:0, mark: 0 (allow-30s-simple-ping-i-1921681234-1111-c-1)
```

"In my computer works (:tm:)" :)

>  Any ideas why there is no pop-up for that?

Maybe the hook is not working, or working differently on the kernel 5.19. I'll test it in a few hours. 
If the hook is working, you should see messages here filtering by ping:
```bash
~ # cat /sys/kernel/debug/tracing/trace_pipe | grep ping
            ping-3655175 [002] d..31 1207647.195302: bpf_trace_printk: sec_socket_sendmsg, type1: 3, proto: 1, fam: 2
```

Before doing these tests, be sure that you restart the daemon. Since you usually hibernate/suspend the computer, ebpf could have stopped working due to the suspend bug. I'd recommend you to install this script -> https://github.com/evilsocket/opensnitch/discussions/834#discussioncomment-5712431 to `/lib/systemd/system-sleep/` with any name and exec permissions.

By the way, you could enable `[x] Debug invalid connections` under Preferences -> Nodes. The GUI should prompt you to deny/allow an "Outgoing connection".

On the other hand, a "modern" ping looks like this:
```bash
# strace -f -e trace=network ping 1.1.1.1 -c 1
socket(AF_INET, SOCK_RAW, IPPROTO_ICMP) = 3
socket(AF_INET6, SOCK_RAW, IPPROTO_ICMPV6) = 4
socket(AF_INET, SOCK_DGRAM, IPPROTO_IP) = 5
connect(5, {sa_family=AF_INET, sin_port=htons(1025), sin_addr=inet_addr("1.1.1.1")}, 16) = 0
```

while the "old" ping:
```bash
socket(AF_INET, SOCK_RAW, IPPROTO_ICMP) = 3
socket(AF_INET6, SOCK_RAW, IPPROTO_ICMPV6) = 4
bind(3, {sa_family=AF_INET, sin_port=htons(0), sin_addr=inet_addr("192.168.1.234")}, 16) = 0
```

Anyway, this needs more investigation.

**Comment by akronym0** [comment]
Ok, thanks.

I did another test after a reboot. This is the output of four terminal windows (deleted log first).

```
❯ sudo ./opensnitchd.946 -rules-path /etc/opensnitchd/rules/
[2023-05-28 15:35:00]  IMP  Starting opensnitch-daemon v1.6.0
[2023-05-28 15:35:00]  INF  Loading rules from /etc/opensnitchd/rules ...
OK: libnetfiler_queue supports nfq_get_uid
OK: libnetfiler_queue supports nfq_get_uid
found /lib/x86_64-linux-gnu/libc.so.6
```

```
❯ sudo cat /sys/kernel/debug/tracing/trace_pipe | grep ping     
<...>-4242    [014] d..31   173.983069: bpf_trace_printk: sec_socket_sendmsg, proto: 1, type: 2, comm: ping
 ping-4242    [014] d..31   174.984998: bpf_trace_printk: sec_socket_sendmsg, type1: 2, proto: 1, fam: 2
 ping-4242    [014] d..31   174.985008: bpf_trace_printk: sec_socket_sendmsg, type1: 2, dport: 0
 ping-4242    [014] d..31   174.985013: bpf_trace_printk: sec_socket_sendmsg, proto: 1, type: 2, comm: ping
```

```
❯ ping -I 192.168.1.111 1.1.1.1 -c 2
PING 1.1.1.1 (1.1.1.1) from 192.168.1.111 : 56(84) bytes of data.
64 bytes from 1.1.1.1: icmp_seq=1 ttl=57 time=6.83 ms
64 bytes from 1.1.1.1: icmp_seq=2 ttl=57 time=7.19 ms

--- 1.1.1.1 ping statistics ---
2 packets transmitted, 2 received, 0% packet loss, time 1002ms
rtt min/avg/max/mdev = 6.827/7.010/7.194/0.183 ms
```


```
❯ tail -f /var/log/opensnitchd.log | grep -i icmp -A 1

[no output]
```

And I noticed this in the log

```
[2023-05-28 15:35:20]  DBG  [eBPF exec event] ppid: 14916759793711452123, pid: 4242, /usr/bin/ping -> [ping -I 192.168.1.111 1.1.1.1 -c 2]
[2023-05-28 15:35:20]  DBG  (2) EBPF-DNS: LookupEvent 272 02000000 01010101000000000000000000000000 312e312e312e310000000000000( ...shortened ...)
[2023-05-28 15:35:20]  DBG  (2) EBPF-DNS: Tracking Resolved Message: 1.1.1.1 -> 1.1.1.1
```
There was nothing else related.

 > Debug invalid connections
 
 This didn't make a difference. The ping command still bypassed OpenSnitch completely.

**Reply by gustavo-iniguez-goya** [reply]
Thank you @akronym0 , I don't want to bother you with more tests, you already have had enough patience :)

tl;dr: no tl;dr today.

After more digging, sometimes when an application is blocked (or not allowed yet) the system sends an ICMP of type DestinationUnreachable(Port) , which could explain why sometimes UDP packets (mainly) are followed by apparently an ICMP packet to the same IP.

For example:
```bash
[2023-05-29 17:29:43]  DBG  new connection udp => 47469:192.168.1.108 -> 91.189.94.4 (ntp.ubuntu.com):123 uid: 102, mark: 0
[2023-05-29 17:29:43]  DBG  [ebpf conn] not in cache, but in execEvents: udp47469192.168.1.10891.189.94.4123, 138170 -> /usr/lib/systemd/systemd-timesyncd
[2023-05-29 17:29:43]  DBG  [ebpf conn] adding item to cache: udp47469192.168.1.10891.189.94.4123
[2023-05-29 17:29:43]  DBG  ✔ /usr/lib/systemd/systemd-timesyncd -> 47469:192.168.1.108 => ntp.ubuntu.com (91.189.94.4):123, mark: 0 (allow-until-restart-simple-usrlibsystemdsystemd-timesyncd)

[2023-05-29 17:29:43]  DBG  new connection icmp => 0:192.168.1.108 -> 91.189.94.4 (ntp.ubuntu.com):0 uid: 4294967295, mark: 0
[2023-05-29 17:29:43]  DBG  netlink socket error: Warning, no message nor error from netlink, or no connections found - 0:192.168.1.108 -> 91.189.94.4:0
```

Another example of an "icmp" packet:
icmp: &{BaseLayer:{Contents:[3 3 55 19 0 0 0 0] Payload:[69 0 0 76 206 38 64 0 53 17 61 174 185 125 190 58 192 168 1 108 0 123 196 99 0 56 92 74 36 2 0 231 0 0 1 228 0 0 0 80 86 23 195 30 232 31 114 39 183 72 67 93 232 31 117 54 33 72 44 39 232 31 117 54 141 191 63 188 232 31 117 54 141 193 78 152]} TypeCode:DestinationUnreachable(Port) Checksum:14099 Id:0 Seq:0}

And just to be sure, comparing it against network captures give us that it's a NTP/UDP packet:
![image](https://github.com/evilsocket/opensnitch/assets/2742953/ee672c38-b640-47dc-9068-f16352add57c)

This could explain why some connections are being attributed to some applications, for example this one to firefox:
>new connection icmp => 0:192.168.1.111 -> 104.22.11.50 (media.printables.com):0 uid:

I can reliably reproduce this problem deleting the rules for `systemd-resolved` and `systemd-timesyncd`, and restarting both daemons.

:point_right: The thing is that we're blocking those ICMP, unconditionally because we're not obtaining the name of the process, thus it doesn't appear on the network captures.

:lady_beetle: I've modified the logic to allow the user to allow/deny these connections by checking `[x] Debug invalid connections`. I'll add it soon.

:bulb: it would be cool to send all the details of a network packet to the server (i.e.: the GUI). Because an "ICMP" packet generally looks like as a echo-ping/echo-reply.

I haven't investigated the kernel part yet, but The Internet says that these ICMP are generated from kernel, and are not associated to any app.

**Reply by akronym0** [reply]
I am happy to test :-)

> The Internet says that these ICMP are generated from kernel, and are not associated to any app.

Thanks. But how about the command

`ping -I 192.168.1.111 1.1.1.1 -c 2`

Why doesn't OpenSnitch intercept that?

**Reply by gustavo-iniguez-goya** [reply]
it needs more analysis, I tested it on kernels 5.8, 5.19, 6.1 and 6.2 and works fine for me.


Are you executing the ping command under a namespace or something like that? firejail, bublewrap.. etc? if you execute it as root does it make any difference?

I've added more logs to the binary ("ICMP NOT found via eBPF" , "ICMP found via eBPF"):

[opensnitchd.946.gz](https://github.com/evilsocket/opensnitch/files/11601495/opensnitchd.946.gz)


[opensnitch.o.gz](https://github.com/evilsocket/opensnitch/files/11600253/opensnitch.o.gz)

The following script dumps the intercepted udp/tcp/icmp packets. If the icmp packets are not intercepted, they should appear there if are of the type 3 or proto == 1 or proto == 58:
[debug-ebpf-maps.sh.txt](https://github.com/evilsocket/opensnitch/files/11600268/debug-ebpf-maps.sh.txt)

`bash debug-ebpf-maps.sh icmp`

(is the same script than this one https://github.com/evilsocket/opensnitch/blob/master/utils/scripts/debug-ebpf-maps.sh , but with icmp support).

**Reply by akronym0** [reply]
It doesn't really change anything. The 946 is more verbose now, but I don't see any log entries with "ICMP NOT found via eBPF" or "ICMP found via eBPF"

The debug script doesn't output anything for icmp.

```
❯ sudo bash debug-ebpf-maps.sh icmp
/usr/sbin/bpftool

Total: 0
```



> Are you executing the ping command under a namespace or something like that? firejail, bublewrap.. etc?

No, just testing in a plain terminal. Root doesn't make a difference.

The only non-standard config in my system is

`GRUB_CMDLINE_LINUX_DEFAULT="ipv6.disable_ipv6=1`

**Reply by gustavo-iniguez-goya** [reply]
and `hping3 -1 1.1.1.1`? see if it prints out the icmp logs `ICMP NOT found via eBPF` || `ICMP found via eBPF`

do you see this log in Debug log level? `[ebpf conn] adding item to cache`

**Comment by akronym0** [comment]
I installed hping3, but that doesn't work on my system. I always get 100% packet loss (with OpenSnitch disabled). I tried many external addresses :-)

```
❯ sudo hping3 -c 3 1.1.1.1
HPING 1.1.1.1 (wlp1s0 1.1.1.1): NO FLAGS are set, 40 headers + 0 data bytes

--- 1.1.1.1 hping statistic ---
3 packets transmitted, 0 packets received, 100% packet loss
round-trip min/avg/max = 0.0/0.0/0.0 ms
```
No, idea why.
